File: internal/stanza/stanza.go
package stanza

import (
	"fmt"
	"jabber/internal/auth"
	"jabber/internal/disco"
	"jabber/internal/message"
	"jabber/internal/presence"
	"jabber/internal/register"
	"jabber/internal/roster"
	"jabber/internal/session"
	"jabber/internal/stream"
	"jabber/internal/xml"
	"log/slog"
	"strings"
)

const (
	nsRegister  = "jabber:iq:register"
	nsBind      = "urn:ietf:params:xml:ns:xmpp-bind"
	nsSession   = "urn:ietf:params:xml:ns:xmpp-session"
	nsRoster    = "jabber:iq:roster"
	nsDiscoInfo = "http://jabber.org/protocol/disco#info"
	nsDiscoItems = "http://jabber.org/protocol/disco#items"
)

// Route dispatches a stanza to the appropriate handler.
func Route(reg session.Registry, s *session.Session, node *xml.XMLNode) session.RouteResult {
	slog.Debug("stanza", "name", node.Name, "ns", node.NS, "state", s.State, "jid", s.BareJID())

	// Pre-auth: only SASL and in-band registration allowed.
	if s.State == session.StateStreamOpened && !s.Authenticated {
		switch node.Name {
		case "auth":
			return auth.HandleSASL(reg, s, node)
		case "iq":
			child := firstChild(node)
			if child != nil && child.NS == nsRegister {
				return register.HandleIQ(reg, s, node)
			}
			// Non-register IQ: stanza error (connection stays open).
			s.SendStanzaError(node, "cancel", "not-allowed")
			return session.RouteOK
		default:
			// Anything else: stream error (connection closes).
			stream.SendError(s, "not-authorized")
			return session.RouteClose
		}
	}

	// Post-auth routing.
	switch node.Name {
	case "iq":
		return handleIQ(reg, s, node)
	case "message":
		if s.State != session.StateSessionActive && s.State != session.StateBound {
			stream.SendError(s, "not-authorized")
			return session.RouteClose
		}
		return message.Handle(reg, s, node)
	case "presence":
		if s.State != session.StateSessionActive && s.State != session.StateBound {
			stream.SendError(s, "not-authorized")
			return session.RouteClose
		}
		return presence.Handle(reg, s, node)
	default:
		stream.SendError(s, "unsupported-stanza-type")
		return session.RouteClose
	}
}

func handleIQ(reg session.Registry, s *session.Session, node *xml.XMLNode) session.RouteResult {
	iqType := node.Attr("type")
	to := node.Attr("to")

	// Route result/error stanzas to their target user.
	if iqType == "result" || iqType == "error" {
		if to != "" && !isServerJID(to, reg.Domain()) {
			targetBare := bareJID(to)
			target := reg.FindByBareJID(targetBare)
			if target != nil {
				setAttr(node, "from", s.FullJID())
				target.Send(node.Serialize())
			}
		}
		return session.RouteOK
	}

	// Dispatch get/set by child element namespace.
	child := firstChild(node)
	childNS := ""
	if child != nil {
		childNS = child.NS
	}

	switch childNS {
	case nsBind:
		return handleBind(reg, s, node)
	case nsSession:
		return handleSessionIQ(s, node)
	case nsRoster:
		if s.State != session.StateSessionActive && s.State != session.StateBound {
			s.SendStanzaError(node, "cancel", "not-allowed")
			return session.RouteOK
		}
		return handleRosterIQ(reg, s, node)
	case nsRegister:
		return register.HandleIQ(reg, s, node)
	case nsDiscoInfo:
		if s.State != session.StateSessionActive && s.State != session.StateBound {
			s.SendStanzaError(node, "cancel", "not-allowed")
			return session.RouteOK
		}
		return disco.HandleInfo(reg, s, node)
	case nsDiscoItems:
		if s.State != session.StateSessionActive && s.State != session.StateBound {
			s.SendStanzaError(node, "cancel", "not-allowed")
			return session.RouteOK
		}
		return disco.HandleItems(reg, s, node)
	default:
		// Unknown namespace: try routing to another user.
		if to != "" && !isServerJID(to, reg.Domain()) &&
			(s.State == session.StateSessionActive || s.State == session.StateBound) {
			targetBare := bareJID(to)
			target := reg.FindByBareJID(targetBare)
			if target != nil {
				setAttr(node, "from", s.FullJID())
				target.Send(node.Serialize())
				return session.RouteOK
			}
		}
		s.SendStanzaError(node, "cancel", "service-unavailable")
		return session.RouteOK
	}
}

// handleBind processes resource binding (RFC 6120 §7).
func handleBind(reg session.Registry, s *session.Session, node *xml.XMLNode) session.RouteResult {
	if s.State != session.StateAuthenticated && s.State != session.StateStreamOpened {
		s.SendStanzaError(node, "cancel", "not-allowed")
		return session.RouteOK
	}

	id := node.Attr("id")

	// Extract requested resource.
	resource := ""
	if bindEl := firstChild(node); bindEl != nil {
		if resEl := bindEl.FindChild("resource"); resEl != nil {
			resource = resEl.Text
		}
	}
	if resource == "" {
		resource = xml.GenerateID(8)
	}

	s.JIDResource = resource
	s.State = session.StateBound
	fullJID := s.FullJID()

	// Register under bare JID; kick any conflicting session.
	if old := reg.Register(s); old != nil {
		slog.Info("session conflict — kicking old session", "jid", old.BareJID())
		stream.SendError(old, "conflict")
		old.Close()
	}

	// Send bind result.
	var b strings.Builder
	fmt.Fprintf(&b, "<iq type='result'")
	if id != "" {
		fmt.Fprintf(&b, " id='%s'", escAttr(id))
	}
	b.WriteString("><bind xmlns='urn:ietf:params:xml:ns:xmpp-bind'>")
	fmt.Fprintf(&b, "<jid>%s</jid>", escText(fullJID))
	b.WriteString("</bind></iq>")
	s.Send(b.String())

	slog.Info("resource bound", "jid", fullJID)
	return session.RouteOK
}

// handleSessionIQ processes session establishment (RFC 3921, deprecated).
func handleSessionIQ(s *session.Session, node *xml.XMLNode) session.RouteResult {
	id := node.Attr("id")
	s.State = session.StateSessionActive

	var b strings.Builder
	b.WriteString("<iq type='result'")
	if id != "" {
		fmt.Fprintf(&b, " id='%s'", escAttr(id))
	}
	b.WriteString("/>")
	s.Send(b.String())

	slog.Info("session established", "jid", s.FullJID())
	return session.RouteOK
}

// handleRosterIQ processes jabber:iq:roster get/set.
func handleRosterIQ(reg session.Registry, s *session.Session, node *xml.XMLNode) session.RouteResult {
	iqType := node.Attr("type")
	id := node.Attr("id")

	if !s.Roster.Loaded {
		s.Roster = roster.Load(reg.DataDir(), s.JIDLocal)
	}

	switch iqType {
	case "get":
		var b strings.Builder
		fmt.Fprintf(&b, "<iq type='result'")
		if id != "" {
			fmt.Fprintf(&b, " id='%s'", escAttr(id))
		}
		fmt.Fprintf(&b, " to='%s'", escAttr(s.FullJID()))
		b.WriteString("><query xmlns='jabber:iq:roster'>")
		for _, item := range s.Roster.Items {
			fmt.Fprintf(&b, "<item jid='%s'", escAttr(item.JID))
			if item.Name != "" {
				fmt.Fprintf(&b, " name='%s'", escAttr(item.Name))
			}
			fmt.Fprintf(&b, " subscription='%s'", item.Subscription)
			if item.AskSubscribe {
				b.WriteString(" ask='subscribe'")
			}
			b.WriteString("/>")
		}
		b.WriteString("</query></iq>")
		s.Send(b.String())

	case "set":
		queryEl := firstChild(node)
		if queryEl == nil {
			s.SendStanzaError(node, "modify", "bad-request")
			return session.RouteOK
		}
		itemEl := queryEl.FindChild("item")
		if itemEl == nil {
			s.SendStanzaError(node, "modify", "bad-request")
			return session.RouteOK
		}
		jid := itemEl.Attr("jid")
		if jid == "" {
			s.SendStanzaError(node, "modify", "bad-request")
			return session.RouteOK
		}
		itemName := itemEl.Attr("name")
		sub := itemEl.Attr("subscription")

		if sub == "remove" {
			s.Roster.Remove(jid)
			_ = roster.Save(reg.DataDir(), s.JIDLocal, &s.Roster)

			// Send result.
			var b strings.Builder
			b.WriteString("<iq type='result'")
			if id != "" {
				fmt.Fprintf(&b, " id='%s'", escAttr(id))
			}
			b.WriteString("/>")
			s.Send(b.String())

			// Roster push with subscription=remove.
			roster.Push(s.Send, s.FullJID(), &roster.Item{
				JID:          jid,
				Subscription: "remove",
			})
		} else {
			// Add or update.
			existing := s.Roster.Find(jid)
			existingSub := "none"
			existingAsk := false
			if existing != nil {
				existingSub = existing.Subscription
				existingAsk = existing.AskSubscribe
			}
			s.Roster.AddOrUpdate(jid, itemName, existingSub, existingAsk)
			if itemName != "" && existing != nil {
				existing.Name = itemName
			}
			_ = roster.Save(reg.DataDir(), s.JIDLocal, &s.Roster)

			var b strings.Builder
			b.WriteString("<iq type='result'")
			if id != "" {
				fmt.Fprintf(&b, " id='%s'", escAttr(id))
			}
			b.WriteString("/>")
			s.Send(b.String())

			updated := s.Roster.Find(jid)
			if updated != nil {
				roster.Push(s.Send, s.FullJID(), updated)
			}
		}

	default:
		s.SendStanzaError(node, "cancel", "feature-not-implemented")
	}

	return session.RouteOK
}

// --- helpers ---

func firstChild(n *xml.XMLNode) *xml.XMLNode {
	if len(n.Children) == 0 {
		return nil
	}
	return n.Children[0]
}

func isServerJID(jid, domain string) bool {
	if jid == "" || jid == domain {
		return true
	}
	return false
}

func bareJID(jid string) string {
	if i := strings.Index(jid, "/"); i >= 0 {
		return jid[:i]
	}
	return jid
}

func setAttr(node *xml.XMLNode, name, value string) {
	for i := range node.Attrs {
		if node.Attrs[i].Name == name {
			node.Attrs[i].Value = value
			return
		}
	}
	node.Attrs = append(node.Attrs, xml.XMLAttr{Name: name, Value: value})
}

func escAttr(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "'", "&apos;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	return s
}

func escText(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	s = strings.ReplaceAll(s, ">", "&gt;")
	return s
}

File: c/src/stanza.c
#include "stanza.h"
#include "session.h"
#include "stream.h"
#include "auth.h"
#include "roster.h"
#include "presence.h"
#include "message.h"
#include "disco.h"
#include "register.h"
#include "config.h"
#include "log.h"
#include "xml.h"
#include "util.h"
#include <string.h>
#include <stdlib.h>
#include <libxml/tree.h>

/* Forward declarations for handlers implemented in other modules */
void session_handle_bind(session_t *s, xmlNodePtr stanza);
void session_handle_session_iq(session_t *s, xmlNodePtr stanza);

/* Forward declarations */
static void handle_iq(session_t *s, xmlNodePtr stanza);

static int is_server_jid(const char *to) {
    /* Check if the JID is addressed to the server (no localpart) */
    if (!to || !*to)
        return 1;
    if (strcmp(to, g_config.domain) == 0)
        return 1;
    return 0;
}

void stanza_route(session_t *s, xmlNodePtr stanza) {
    const char *name = (const char *)stanza->name;
    const char *ns = stanza->ns ? (const char *)stanza->ns->href : "";

    log_write(LOG_DEBUG, "Stanza received on fd %d: <%s> ns='%s' state=%d presence_stanza=%p",
              s->fd, name, ns, s->state, (void *)s->presence_stanza);

    /* Pre-auth: only SASL and in-band registration allowed */
    if (s->state == STATE_STREAM_OPENED && !s->authenticated) {
        if (strcmp(name, "auth") == 0 &&
            strcmp(ns, "urn:ietf:params:xml:ns:xmpp-sasl") == 0) {
            auth_handle_sasl(s, stanza);
        } else if (strcmp(name, "iq") == 0) {
            /* Allow registration IQs only */
            xmlNodePtr child = stanza->children;
            while (child && child->type != XML_ELEMENT_NODE)
                child = child->next;
            const char *cns = (child && child->ns && child->ns->href)
                              ? (const char *)child->ns->href : "";
            if (strcmp(cns, "jabber:iq:register") == 0)
                register_handle_iq(s, stanza);
            else
                stanza_send_error(s, stanza, "cancel", "not-allowed");
        } else {
            stream_send_error(s, "not-authorized");
        }
        return;
    }

    /* Post-auth: IQ for bind/session, then full routing once active */
    if (strcmp(name, "iq") == 0) {
        handle_iq(s, stanza);
    } else if (strcmp(name, "message") == 0) {
        if (s->state != STATE_SESSION_ACTIVE && s->state != STATE_BOUND) {
            stream_send_error(s, "not-authorized");
            return;
        }
        handle_message(s, stanza);
    } else if (strcmp(name, "presence") == 0) {
        if (s->state != STATE_SESSION_ACTIVE && s->state != STATE_BOUND) {
            stream_send_error(s, "not-authorized");
            return;
        }
        handle_presence(s, stanza);
    } else {
        stream_send_error(s, "unsupported-stanza-type");
    }
}

static void handle_iq(session_t *s, xmlNodePtr stanza) {
    xmlChar *type_attr = xmlGetProp(stanza, (const xmlChar *)"type");
    xmlChar *to_attr   = xmlGetProp(stanza, (const xmlChar *)"to");
    const char *type = type_attr ? (const char *)type_attr : "";
    const char *to   = to_attr ? (const char *)to_attr : "";

    /* result/error: route to target user if online, else drop */
    if (strcmp(type, "result") == 0 || strcmp(type, "error") == 0) {
        if (to[0] && !is_server_jid(to)) {
            /* Route to target user */
            char local[256], domain[256], resource[256];
            jid_parse(to, local, sizeof(local), domain, sizeof(domain),
                      resource, sizeof(resource));
            char bare[512];
            jid_bare(local, domain, bare, sizeof(bare));
            session_t *target = session_find_by_jid(bare);
            if (target) {
                /* Set from to sender's full JID */
                char from_jid[768];
                jid_full(s->jid_local, s->jid_domain, s->jid_resource,
                         from_jid, sizeof(from_jid));
                xmlSetProp(stanza, (const xmlChar *)"from",
                           (const xmlChar *)from_jid);
                stanza_send(target, stanza);
            }
        }
        if (type_attr) xmlFree(type_attr);
        if (to_attr) xmlFree(to_attr);
        return;
    }

    /* get/set: identify child element namespace for dispatch */
    xmlNodePtr child = stanza->children;
    while (child && child->type != XML_ELEMENT_NODE)
        child = child->next;

    const char *child_ns = "";
    if (child && child->ns && child->ns->href)
        child_ns = (const char *)child->ns->href;

    if (strcmp(child_ns, "urn:ietf:params:xml:ns:xmpp-bind") == 0) {
        session_handle_bind(s, stanza);
    } else if (strcmp(child_ns, "urn:ietf:params:xml:ns:xmpp-session") == 0) {
        session_handle_session_iq(s, stanza);
    } else if (strcmp(child_ns, "jabber:iq:roster") == 0) {
        if (s->state != STATE_SESSION_ACTIVE && s->state != STATE_BOUND) {
            stanza_send_error(s, stanza, "cancel", "not-allowed");
        } else {
            roster_handle_iq(s, stanza);
        }
    } else if (strcmp(child_ns, "http://jabber.org/protocol/disco#info") == 0) {
        if (s->state != STATE_SESSION_ACTIVE && s->state != STATE_BOUND) {
            stanza_send_error(s, stanza, "cancel", "not-allowed");
        } else {
            disco_handle_info(s, stanza);
        }
    } else if (strcmp(child_ns, "http://jabber.org/protocol/disco#items") == 0) {
        if (s->state != STATE_SESSION_ACTIVE && s->state != STATE_BOUND) {
            stanza_send_error(s, stanza, "cancel", "not-allowed");
        } else {
            disco_handle_items(s, stanza);
        }
    } else if (strcmp(child_ns, "jabber:iq:register") == 0) {
        register_handle_iq(s, stanza);
    } else {
        /* Unknown namespace: if addressed to another user, route; else error */
        if (to[0] && !is_server_jid(to) &&
            (s->state == STATE_SESSION_ACTIVE || s->state == STATE_BOUND)) {
            char local[256], domain[256], resource[256];
            jid_parse(to, local, sizeof(local), domain, sizeof(domain),
                      resource, sizeof(resource));
            char bare[512];
            jid_bare(local, domain, bare, sizeof(bare));
            session_t *target = session_find_by_jid(bare);
            if (target) {
                char from_jid[768];
                jid_full(s->jid_local, s->jid_domain, s->jid_resource,
                         from_jid, sizeof(from_jid));
                xmlSetProp(stanza, (const xmlChar *)"from",
                           (const xmlChar *)from_jid);
                stanza_send(target, stanza);
            } else {
                stanza_send_error(s, stanza, "cancel", "service-unavailable");
            }
        } else {
            stanza_send_error(s, stanza, "cancel", "service-unavailable");
        }
    }

    if (type_attr) xmlFree(type_attr);
    if (to_attr) xmlFree(to_attr);
}

/* --- Serialization utilities --- */

char *stanza_serialize(xmlNodePtr node, size_t *out_len) {
    xmlBufferPtr buf = xmlBufferCreate();
    if (!buf)
        return NULL;

    int len = xmlNodeDump(buf, NULL, node, 0, 0);
    if (len < 0) {
        xmlBufferFree(buf);
        return NULL;
    }

    char *result = malloc((size_t)len + 1);
    if (!result) {
        xmlBufferFree(buf);
        return NULL;
    }

    memcpy(result, xmlBufferContent(buf), (size_t)len);
    result[len] = '\0';
    if (out_len)
        *out_len = (size_t)len;

    xmlBufferFree(buf);
    return result;
}

void stanza_send(session_t *s, xmlNodePtr node) {
    size_t len;
    char *xml = stanza_serialize(node, &len);
    if (xml) {
        session_write(s, xml, len);
        free(xml);
    }
}

void stanza_send_error(session_t *s, xmlNodePtr original,
                       const char *error_type, const char *condition)
{
    const char *tag = (const char *)original->name;
    xmlChar *id = xmlGetProp(original, (const xmlChar *)"id");

    xmlNodePtr err = xmlNewNode(NULL, (const xmlChar *)tag);
    xmlNewProp(err, (const xmlChar *)"type", (const xmlChar *)"error");
    if (id) {
        xmlNewProp(err, (const xmlChar *)"id", id);
        xmlFree(id);
    }
    xmlNewProp(err, (const xmlChar *)"from", (const xmlChar *)g_config.domain);

    char full_jid[768];
    jid_full(s->jid_local, s->jid_domain, s->jid_resource, full_jid, sizeof(full_jid));
    xmlNewProp(err, (const xmlChar *)"to", (const xmlChar *)full_jid);

    xmlNodePtr error_el = xmlNewChild(err, NULL, (const xmlChar *)"error", NULL);
    xmlNewProp(error_el, (const xmlChar *)"type", (const xmlChar *)error_type);

    xmlNodePtr cond = xmlNewChild(error_el, NULL, (const xmlChar *)condition, NULL);
    xmlNsPtr ns = xmlNewNs(cond,
        (const xmlChar *)"urn:ietf:params:xml:ns:xmpp-stanzas", NULL);
    xmlSetNs(cond, ns);

    stanza_send(s, err);
    xmlFreeNode(err);
}
